src/aind_data_asset_indexer/utils.py

```diff
@@ -0,0 +1,300 @@
+    a valid json file.
+    Parameters
+    ----------
+    s3_client : S3Client
+    bucket : str
+    object_key : str
+
+    Returns
+    -------
+    Optional[dict]
+
+    """
+    result = s3_client.get_object(Bucket=bucket, Key=object_key)
+    try:
+        content_string = result["Body"].read().decode(encoding="utf-8")
+        return json.loads(content_string)
+    except JSONDecodeError:
+        return None
+
+
+def build_metadata_record_from_prefix(
+    bucket: str,
+    metadata_nd_overwrite: bool,
+    prefix: str,
+    s3_client: S3Client,
+) -> Optional[str]:
+    """
+    For a given bucket and prefix, this method will return a JSON string
+    representation of a Metadata record. If metadata_nd_overwrite is True or
+    a metadata.nd.json file does not exist, then a Metadata record will be
+    constructed from any core schema json files found under the prefix.
+    Otherwise, the method will return the metadata.nd.json file found in S3
+    as a JSON string if it is valid json. If not valid json, then it will
+    return None.
+    Parameters
+    ----------
+    bucket : str
+    metadata_nd_overwrite : bool
+    prefix : str
+    s3_client : S3Client
+
+    Returns
+    -------
+    Optional[str]
+      The Metadata record as a json string. Will return None if
+      metadata_nd_overwrite is set to false, there is a metadata.nd.json file,
+      and the file is corrupt.
+
+    """
+    stripped_prefix = prefix[:-1] if prefix.endswith("/") else prefix
+    metadata_nd_file_key = stripped_prefix + "/" + Metadata.default_filename()
+    does_metadata_nd_file_exist = does_s3_object_exist(
+        s3_client=s3_client, bucket=bucket, key=metadata_nd_file_key
+    )
+    if metadata_nd_overwrite or not does_metadata_nd_file_exist:
+        file_keys = [
+            stripped_prefix + "/" + file_name
+            for file_name in core_schema_file_names
+        ]
+        s3_file_responses = get_dict_of_file_info(
+            s3_client=s3_client, bucket=bucket, keys=file_keys
+        )
+        # Strip the trailing slash from the prefix
+        metadata_dict = {
+            "name": stripped_prefix,
+            "location": f"s3://{bucket}/{stripped_prefix}",
+        }
+        for object_key, response_data in s3_file_responses.items():
+            if response_data is not None:
+                field_name = object_key.split("/")[-1].replace(".json", "")
+                json_contents = download_json_file_from_s3(
+                    s3_client=s3_client, bucket=bucket, object_key=object_key
+                )
+                if json_contents is not None:
+                    # Old version of pycharm highlights a warning since
+                    # it doesn't know check above ensures json_contents is not
+                    # None
+                    # noinspection PyTypeChecker
+                    is_corrupt = is_dict_corrupt(input_dict=json_contents)
```

Comment:
If it's corrupt, it returns metadata_dict without the data -- is it supposed to return None ?

Reply:
It's set to None as default. For example, it will download the `processing.json` file. If it is not corrupt, it will add something like `metadata_dict["processing"] = json.load("processing.json")`. Otherwise, it will move forward to the next file. The metadata_dict is created above outside the for loop with its required fields.